src/optimize_1q_gates.rs

```diff
@@ -0,0 +1,102 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+const PI: f64 = std::f64::consts::PI;
+
+//     Return a triple theta, phi, lambda for the product.
+//
+//         u3(theta, phi, lambda)
+//            = u3(theta1, phi1, lambda1).u3(theta2, phi2, lambda2)
+//            = Rz(phi1).Ry(theta1).Rz(lambda1+phi2).Ry(theta2).Rz(lambda2)
+//            = Rz(phi1).Rz(phi').Ry(theta').Rz(lambda').Rz(lambda2)
+//            = u3(theta', phi1 + phi', lambda2 + lambda')
+//
+//         Return theta, phi, lambda.
```

Comment:
Using `///` will treat this comment as a docstring in rust. The `pyfunction` macro knows to convert rust docstrings into the `__doc__` attribute of the generated python function. This way we'll have a docstring for the python function, it won't be published with the rendered docs since we don't expose this function publicly but setting the docstring will enable things to inspect it and some IDEs to show it in python.

```suggestion
///     Return a triple theta, phi, lambda for the product.
///
///         u3(theta, phi, lambda)
///            = u3(theta1, phi1, lambda1).u3(theta2, phi2, lambda2)
///            = Rz(phi1).Ry(theta1).Rz(lambda1+phi2).Ry(theta2).Rz(lambda2)
///            = Rz(phi1).Rz(phi').Ry(theta').Rz(lambda').Rz(lambda2)
///            = u3(theta', phi1 + phi', lambda2 + lambda')
///
///         Return theta, phi, lambda.
```